internal/download/get.go

```diff
@@ -31,81 +34,115 @@ func init() {
+	} else if res.StatusCode != http.StatusOK {
+		logging.Debug("Could not grab url due to statuscode: %d", res.StatusCode)
 	} else {
-		total, err = strconv.Atoi(length)
+		lengthInt, err := strconv.Atoi(res.Header.Get("Content-Length"))
 		if err != nil {
-			logging.Debug("Content-length: %v", length)
-			return nil, failures.FailInput.Wrap(err)
+			logging.Debug("Could not grab content-length: %v", err)
+		} else {
+			length = int64(lengthInt)
 		}
 	}
 
-	bar := progress.AddByteProgressBar(int64(total))
-
-	src := resp.Body
-	var dst bytes.Buffer
+	// Close early cause we're just looking at the header
+	// Yes normally you'd use a HEAD for this, but S3 presigned URLs don't support HEAD requests
+	res.Body.Close()
 
-	src = bar.ProxyReader(resp.Body)
+	// Record progress
+	bar := progress.AddByteProgressBar(length)
+	cb := func(length int) {
+		if !bar.Completed() {
+			// Failsafe, so we don't get blocked by a progressbar
+			bar.IncrBy(length)
+		}
+	}
 
-	_, err = io.Copy(&dst, src)
+	// Prepare result recorder
+	b := []byte{}
+	w := NewWriteAtBuffer(b, cb)
+
+	dl := s3manager.NewDownloader(sess)
+	dl.RequestOptions = append(dl.RequestOptions, func(r *request.Request) {
+		r.Handlers.Build.PushBack(func(r *request.Request) {
+			// Work around AWS rewriting our query in the wrong order, causing signing to fail
+			r.HTTPRequest.URL.RawQuery = url.RawQuery
+		})
+	})
+	_, err = dl.Download(w,
+		&s3.GetObjectInput{
+			Bucket: aws.String(s3m.Bucket),
+			Key:    aws.String(s3m.Key),
+		})
 	if err != nil {
-		return nil, failures.FailInput.Wrap(err)
+		return nil, locale.WrapError(err, "err_dl_s3", "Downloading failed due to underlying S3 error: {{.V0}}.", err.Error())
 	}
 
-	if !bar.Completed() {
-		// Failsafe, so we don't get blocked by a progressbar
-		bar.IncrBy(total)
-	}
+	bar.Abort(true) // ensure we don't get stuck on an incomplete bar
 
-	return dst.Bytes(), nil
+	return w.Bytes(), nil
 }
 
-func _testHTTPGetWithProgress(url string, progress *progress.Progress) ([]byte, *failures.Failure) {
-	return _testHTTPGet(url)
+type s3Meta struct {
+	Bucket string
+	Region string
+	Key    string
 }
 
-// _testHTTPGet is used when in tests, this cannot be in the test itself as that would limit it to only that one test
-func _testHTTPGet(url string) ([]byte, *failures.Failure) {
-	path := strings.Replace(url, constants.APIArtifactURL, "", 1)
+func parseS3URL(url *url.URL) s3Meta {
+	r := s3Meta{Key: url.Path}
+	domain := strings.SplitN(url.Host, ".", 5)
+	if strings.HasSuffix(url.Host, ".s3.amazonaws.com") { // https://bucket-name.s3.amazonaws.com/key-name
+		r.Bucket = domain[0]
+		r.Region = "us-east-1"
+	} else if domain[1] == "s3" && domain[3] == "amazonaws" { // https://bucket-name.s3.Region.amazonaws.com/key-name
+		r.Bucket = domain[0]
+		r.Region = domain[2]
+	} else { // https://s3.Region.amazonaws.com/bucket-name/key-name
+		r.Bucket = strings.SplitN(url.Path, "/", 1)[0]
+		r.Region = domain[1]
+	}
+	return r
+}
```

Comment:
strong gritty nit:  The index bounds should be checked before access and error conditions should be returned. The improved error message content would provide a trail to follow in the (yes, unlikely) case an incompatible URL is passed in.

Reply:
I falsely assumed `SplitN` would ensure the length of the resulting slice. Will fix.

Reply:
Just for reference, here is what I came up with trialing stuff yesterday (the logic may be off from what you intend, so this isn't necessarily meant to be a drop-in replacement): https://play.golang.org/p/r-qgMHUEmbi